### ISSUE TITLE
测试集预测结果样式和自定义图片的结果样式

### ISSUE THREAD

**Issue by cactusgame** [dailenson/SDT#35]
非常感谢分享了这么好的项目！
初步尝试使用后，有2个问题希望请教一下

### 问题一：使用源码测试集的结果
下图
- 第一行由模型预测产生
- 第二行是测试集内`coordinates`数据。
我使用的模型是github上预训练好的模型执行`test.py`，`sample_size`是500。尝试预测了很多次，预测的文字style看起来和测试集有一定差距，效果看起来不如论文所述。请问可能是什么原因造成的？

![image](https://github.com/dailenson/SDT/assets/5027167/8671e98e-2840-4fc0-9af4-10c0b0b39898)

### 问题二：使用自定义图片的结果
我使用电脑自带的”黑体“作为新的writer，并同时修改了writer_dict等配置，模型未进行微调（和上面的模型一样），期望模型的预测结果接近“黑体”。但结果如下，和“黑体”相差很大

![image](https://github.com/dailenson/SDT/assets/5027167/77c85880-d0b1-42c7-b838-944a98b651c2)

在论文中，我看到如下两段比较困惑：
- 下面这段提到，你们使用了writer的数据的80%进行训练，这是意味着我需要用“黑体”数据进行微调才可以么？
> 【A4】We randomly select 80% of the entire dataset as the training set, and the remaining 20% as the test set.

- 但下面这段话，没有具体说模型是否用`some user`的数据训练或微调过。
> 【A6】As shown in Figure 11 (a), by taking some pictures of the user’s handwriting set as the style reference, our method can readily generate the target-stylized online characters.

所以，第二个问题是，我预期模型可以输出“黑体”样式，那么是否需要用“黑体”进行微调？如果需要微调，大概需要准备多少数据？是否有最佳实践可供参考？

**Comment by dailenson** [comment]
感谢对我们工作的关注~关于``第一个问题``，可以先测一下生成结果的DTW指标，如果DTW有问题就是性能没有复现出来。``第二个问题``，我们的工作是生成手写体的，黑体跟手写体还有有一些差异的，建议自己手写一些字然后输入到模型中去生成，这个过程是不需要finetune的，``具体的教程``近期我会出一个``放在置顶issue``上。``A4``和``A6``可以不用管，``A4``是生成离线的手写中文，是另外一个任务，``A6``主要是讨论模型的输入选择轨迹模态还是离线模态比较好。

**Comment by cactusgame** [comment]
> 感谢对我们工作的关注~关于`第一个问题`，可以先测一下生成结果的DTW指标，如果DTW有问题就是性能没有复现出来。`第二个问题`，我们的工作是生成手写体的，黑体跟手写体还有有一些差异的，建议自己手写一些字然后输入到模型中去生成，这个过程是不需要finetune的，`具体的教程`近期我会出一个`放在置顶issue`上。`A4`和`A6`可以不用管，`A4`是生成离线的手写中文，是另外一个任务，`A6`主要是讨论模型的输入选择轨迹模态还是离线模态比较好。

非常感谢回复。

- 问题1
dtw显示`the avg fast_norm_len_dtw is 43.8289909362793`, 这次结果如下。虽然和测试集不是特别像，但能看出style是相近的。
![image](https://github.com/dailenson/SDT/assets/5027167/bd9d622c-3245-4800-ab74-542ae58ce7ad)

- 问题2
我试过多个不同的机打体，包括机打手写体（如翩翩体），预测出的结果都和下图类似，和输入的style_image相差甚远。我正在尝试照片手写，有进展了我会继续同步。

![image](https://github.com/dailenson/SDT/assets/5027167/ee52b5a8-600c-4e88-a68b-412886990cda)

**Comment by cactusgame** [comment]
> `第二个问题`，我们的工作是生成手写体的，黑体跟手写体还有有一些差异的，建议自己手写一些字然后输入到模型中去生成，

我似乎GET到了，用机打体生成的图片，和手写体最大的不同是“笔触粗细”。SDT似乎只能针对“细线”字迹进行style的模仿。不知道我的理解对不对。

**Comment by cactusgame** [comment]
@dailenson 麻烦问一下，test.py运行时，默认每次只从writer的字迹中取`15`个作为style参考，`15`是如何确定的？是不是这里采样越多输入给模型，style就会和writer的字迹越像？

**Comment by dailenson** [comment]
> on 麻烦问一下，test.py运行时，默认每次只从writer的字迹中取`15`个作为style参考，`15`是如何确定的？是不是这里采样越多输入给模型，style就会和writer的字迹越像？



> > 感谢对我们工作的关注~关于`第一个问题`，可以先测一下生成结果的DTW指标，如果DTW有问题就是性能没有复现出来。`第二个问题`，我们的工作是生成手写体的，黑体跟手写体还有有一些差异的，建议自己手写一些字然后输入到模型中去生成，这个过程是不需要finetune的，`具体的教程`近期我会出一个`放在置顶issue`上。`A4`和`A6`可以不用管，`A4`是生成离线的手写中文，是另外一个任务，`A6`主要是讨论模型的输入选择轨迹模态还是离线模态比较好。
> 
> 非常感谢回复。
> 
> * 问题1
>   dtw显示`the avg fast_norm_len_dtw is 43.8289909362793`, 这次结果如下。虽然和测试集不是特别像，但能看出style是相近的。
>   ![image](https://private-user-images.githubusercontent.com/5027167/292448900-bd9d622c-3245-4800-ab74-542ae58ce7ad.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDMzOTcxMDEsIm5iZiI6MTcwMzM5NjgwMSwicGF0aCI6Ii81MDI3MTY3LzI5MjQ0ODkwMC1iZDlkNjIyYy0zMjQ1LTQ4MDAtYWI3NC01NDJhZTU4Y2U3YWQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMjRUMDU0NjQxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTU3ZDU4ZmMxNzJiMzUzYjk0ZDZkOTYwMzg1YzJiOGQ3MGQ1MDc1NDAyYWZlZGNhYWQ3ODAyNGZkYjJiMzQ1ZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.g24yMy03WvxYW933U0c9zs5zVuun5oxHUrdGPi21Cow)
> * 问题2
>   我试过多个不同的机打体，包括机打手写体（如翩翩体），预测出的结果都和下图类似，和输入的style_image相差甚远。我正在尝试照片手写，有进展了我会继续同步。
> 
> ![image](https://private-user-images.githubusercontent.com/5027167/292449873-ee52b5a8-600c-4e88-a68b-412886990cda.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDMzOTcxMDEsIm5iZiI6MTcwMzM5NjgwMSwicGF0aCI6Ii81MDI3MTY3LzI5MjQ0OTg3My1lZTUyYjVhOC02MDBjLTRlODgtYTY4Yi00MTI4ODY5OTBjZGEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMjRUMDU0NjQxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODE4YzdkZDBkZTQwYzg4MTExMDUzNDQ3OTIzNWFlZWYzZWUyOGZiMTg2OWUwMDQ2YTBkZDBkM2YyZTk5ZTIyMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.b2JLuL-hrnhWGaRlTMlMqbaJ_4t_spwsMNH7fe8IhBQ)

正常情况下，dtw指标对结果应该是0.8左右，你这边可能得检查下是否运行过程中是否有错误。

**Comment by dailenson** [comment]
> @dailenson 麻烦问一下，test.py运行时，默认每次只从writer的字迹中取`15`个作为style参考，`15`是如何确定的？是不是这里采样越多输入给模型，style就会和writer的字迹越像？

15是一个比较折中的数值，可以得到比较好的性能，并且这个数量也不算太多。是的，一般来说给的样本参考越多，模型的性能会更好。

**Comment by dailenson** [comment]
> > 感谢对我们工作的关注~关于`第一个问题`，可以先测一下生成结果的DTW指标，如果DTW有问题就是性能没有复现出来。`第二个问题`，我们的工作是生成手写体的，黑体跟手写体还有有一些差异的，建议自己手写一些字然后输入到模型中去生成，这个过程是不需要finetune的，`具体的教程`近期我会出一个`放在置顶issue`上。`A4`和`A6`可以不用管，`A4`是生成离线的手写中文，是另外一个任务，`A6`主要是讨论模型的输入选择轨迹模态还是离线模态比较好。
> 
> 非常感谢回复。
> 
> * 问题1
>   dtw显示`the avg fast_norm_len_dtw is 43.8289909362793`, 这次结果如下。虽然和测试集不是特别像，但能看出style是相近的。
>   ![image](https://private-user-images.githubusercontent.com/5027167/292448900-bd9d622c-3245-4800-ab74-542ae58ce7ad.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDMzOTcyOTksIm5iZiI6MTcwMzM5Njk5OSwicGF0aCI6Ii81MDI3MTY3LzI5MjQ0ODkwMC1iZDlkNjIyYy0zMjQ1LTQ4MDAtYWI3NC01NDJhZTU4Y2U3YWQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMjRUMDU0OTU5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTI0MmRmZDkzZWMxYmY2ZjgyYWQ3ODBlM2Q1MTgyZjJmNTY1OWVlNzY2NWViOWQ5MjlmZjZlYmI5ODg3OTc2YiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.u_g07aKPKgNW92t9njWMrzV4KatOXGdce_gH6xXKf1o)
> * 问题2
>   我试过多个不同的机打体，包括机打手写体（如翩翩体），预测出的结果都和下图类似，和输入的style_image相差甚远。我正在尝试照片手写，有进展了我会继续同步。
> 
> ![image](https://private-user-images.githubusercontent.com/5027167/292449873-ee52b5a8-600c-4e88-a68b-412886990cda.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDMzOTcyOTksIm5iZiI6MTcwMzM5Njk5OSwicGF0aCI6Ii81MDI3MTY3LzI5MjQ0OTg3My1lZTUyYjVhOC02MDBjLTRlODgtYTY4Yi00MTI4ODY5OTBjZGEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIyNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMjRUMDU0OTU5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODdjNTc3NDlhYTNiYTlmNTE4OTAwYjA4MDFmNmEzMjBlZjE2NzM0NzhhOWEyZWMwZDBjNjkyNmE4MWM2OTNhNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.qxHYz1KuGV0TPerOgWvCJCMgyPxuVjRA0SA5qYsUKLE)

正常运行的情况下``dtw``指标的结果应该是``0.8``左右。你复现出的结果差的有点多，建议再仔细看下readme的教程，看看运行过程中是否有错误。

**Comment by cactusgame** [comment]
> 正常运行的情况下`dtw`指标的结果应该是`0.8`左右。你复现出的结果差的有点多，建议再仔细看下readme的教程，看看运行过程中是否有错误。

我一直在用readme的这条命令进行测试，日志里没有发现错误信息
```
 python test.py --pretrained_model checkpoint_path --store_type online --sample_size 500 --dir Generated/Chinese
```

我的理解，这里的`sample_size`, 就算用100以下甚至更小的值，DTW都应该像你所说的比较小才对，是不是？

**Comment by cactusgame** [comment]
非常感谢，我已经可以复现0.8左右的dtw。上面的错误是由于我稍微修改了一点代码，调用了utils里的coords_render方法，这个方法内部会改变入参的值

![image](https://github.com/dailenson/SDT/assets/5027167/1f3c174b-fd14-4be2-8442-2033f7460123)

> 正常运行的情况下`dtw`指标的结果应该是`0.8`左右。你复现出的结果差的有点多，建议再仔细看下readme的教程，看看运行过程中是否有错误。

**Comment by cactusgame** [comment]
补充一下，0.8左右的dtw是使用test集复现的。但使用我自己画的图片，或者使用字体生成的图片，dtw在4左右。

> 非常感谢，我已经可以复现0.8左右的dtw。上面的错误是由于我稍微修改了一点代码，调用了utils里的coords_render方法，这个方法内部会改变入参的值

**Comment by cactusgame** [comment]
有个疑问，得出dtw 0.8的测试集的`writer`,同一个writer的字迹是否在训练集中出现过呢？我看train style 和test style中writer id有重复的（指0,1,2,3这些数字），但writer fname又不一样

> `A4`和`A6`可以不用管，`A4`是生成离线的手写中文，是另外一个任务，`A6`主要是讨论模型的输入选择轨迹模态还是离线模态比较好。

**Comment by dailenson** [comment]
> 有个疑问，得出dtw 0.8的测试集的`writer`,同一个writer的字迹是否在训练集中出现过呢？我看train style 和test style中writer id有重复的（指0,1,2,3这些数字），但writer fname又不一样。所以测试集中的writer id，和train中的writer id，是否代表同一个writer？
> 
> > `A4`和`A6`可以不用管，`A4`是生成离线的手写中文，是另外一个任务，`A6`主要是讨论模型的输入选择轨迹模态还是离线模态比较好。

训练集跟测试集的书写者是不相交的，在训练过程中从来没见过测试集中的书写者。你所说的writer id是后处理过的，训练集中的0和测试集中的0并不代表一个东西。数据集的划分详情可以进一步看下我们的论文，上面写的很清楚。

**Comment by cactusgame** [comment]
非常感谢，这下清晰多了
> 训练集跟测试集的书写者是不相交的，在训练过程中从来没见过测试集中的书写者。你所说的writer id是后处理过的，训练集中的0和测试集中的0并不代表同一个书写者。数据集的划分详情可以进一步看下我们的论文，上面写的很清楚。

**Comment by aceliuchanghong** [comment]
您好,
我有一些关于SDT的问题想要咨询一下,不知道可以加一下微信吗?
我对AI生成字体非常感兴趣,不知道可以付费咨询一下吗
我的微信号是:sisconsavior
